File: src/observability/schema.ts
/* src/observability/schema.ts — minimal JSON Schema validator for the
 * Enchanter event-bus wire format.
 *
 * Loads docs/event-schema.json at module init and exposes:
 *   - validate(event):    inbound event (Bridge.enqueue → JSONL line)
 *   - validateCommand(c): outbound control command (TcpControlSink → runtime)
 *
 * NOT a full draft 2020-12 implementation. Hand-rolled walker that supports
 * exactly the keywords our schema uses: type, properties, required, oneOf,
 * enum, const, minimum, $ref, additionalProperties. Intentionally tight so
 * "no new top-level dep" is achievable.
 *
 * Failure shape is shared across both sides of the wire — the TS validator
 * and the Rust validator return the same { ok | reason | path } structure
 * so error messages read consistently in mixed-language logs.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ValidationResult =
  | { ok: true }
  | { ok: false; reason: string; path: string[] };

type SchemaNode = Record<string, unknown>;

// ---------------------------------------------------------------------------
// Schema load — resolved relative to this source file so packaged dist works
// ---------------------------------------------------------------------------

function loadSchema(): SchemaNode {
  // import.meta.url points at .ts in dev (tsx/vitest) and .js in dist.
  const here = fileURLToPath(import.meta.url);
  const candidates = [
    // src layout: src/observability/ -> ../../docs/event-schema.json
    path.resolve(path.dirname(here), '..', '..', 'docs', 'event-schema.json'),
    // dist layout: dist/observability/ -> ../../docs/event-schema.json (also valid)
    path.resolve(path.dirname(here), '..', '..', 'docs', 'event-schema.json'),
  ];
  for (const candidate of candidates) {
    try {
      const text = fs.readFileSync(candidate, 'utf8');
      return JSON.parse(text) as SchemaNode;
    } catch {
      /* try next */
    }
  }
  throw new Error(
    `[schema] event-schema.json not found; searched: ${candidates.join(', ')}`,
  );
}

const SCHEMA = loadSchema();
const DEFINITIONS = (SCHEMA.definitions as Record<string, SchemaNode> | undefined) ?? {};

// ---------------------------------------------------------------------------
// Walker
// ---------------------------------------------------------------------------

function deref(node: SchemaNode): SchemaNode {
  const ref = node.$ref;
  if (typeof ref !== 'string') return node;
  // We only support the form "#/definitions/<name>".
  const m = /^#\/definitions\/([^/]+)$/.exec(ref);
  if (!m) return node;
  const name = m[1];
  if (name === undefined) return node;
  const target = DEFINITIONS[name];
  return target ?? node;
}

function jsonType(value: unknown): string {
  if (value === null) return 'null';
  if (Array.isArray(value)) return 'array';
  return typeof value;
}

function typeMatches(declared: string, actual: string): boolean {
  if (declared === 'integer') return actual === 'number';
  return declared === actual;
}

function fail(reason: string, path: string[]): ValidationResult {
  return { ok: false, reason, path };
}

/** Heuristic: does this branch's `type` discriminator (a const or an enum on
 *  the `type` property) match the `type` field of the candidate object?
 *  Used when surfacing oneOf failures so the most-relevant branch's reason
 *  wins over an unrelated branch's `const` mismatch. */
function branchTypeMatches(branch: SchemaNode, value: unknown): boolean {
  if (typeof value !== 'object' || value === null) return false;
  const obj = value as Record<string, unknown>;
  const t = obj.type;
  if (typeof t !== 'string') return false;
  const properties = branch.properties as Record<string, SchemaNode> | undefined;
  if (!properties) return false;
  const typeSchema = properties.type;
  if (!typeSchema) return false;
  if (Object.prototype.hasOwnProperty.call(typeSchema, 'const')) {
    return typeSchema.const === t;
  }
  if (Array.isArray(typeSchema.enum)) {
    return typeSchema.enum.includes(t as never);
  }
  return false;
}

/** Is this branch a strict-discriminator branch (its `type` property pins a
 *  single `const`)? Used to enforce no-fallthrough: when the input's `type`
 *  matches a strict branch's const, only that branch may match — we will not
 *  fall through to the generic permissive variant. Mirrors the Rust
 *  serde(tag = "type") enum dispatch in inspector/src/event.rs, which has no
 *  #[serde(other)] catch-all. */
function branchIsStrictConst(branch: SchemaNode): boolean {
  const properties = branch.properties as Record<string, SchemaNode> | undefined;
  if (!properties) return false;
  const typeSchema = properties.type;
  if (!typeSchema) return false;
  return Object.prototype.hasOwnProperty.call(typeSchema, 'const');
}

function ok(): ValidationResult {
  return { ok: true };
}

function check(node: SchemaNode, value: unknown, path: string[]): ValidationResult {
  const schema = deref(node);

  // const
  if (Object.prototype.hasOwnProperty.call(schema, 'const')) {
    if (value !== schema.const) {
      return fail(`expected const ${JSON.stringify(schema.const)}, got ${JSON.stringify(value)}`, path);
    }
  }

  // enum
  if (Array.isArray(schema.enum)) {
    if (!schema.enum.includes(value as never)) {
      return fail(`value ${JSON.stringify(value)} not in enum`, path);
    }
  }

  // type
  if (typeof schema.type === 'string') {
    const actual = jsonType(value);
    if (!typeMatches(schema.type, actual)) {
      return fail(`expected type ${schema.type}, got ${actual}`, path);
    }
  }

  // minimum (numbers only)
  if (typeof schema.minimum === 'number' && typeof value === 'number') {
    if (value < schema.minimum) {
      return fail(`value ${value} below minimum ${schema.minimum}`, path);
    }
  }

  // object structure
  if (jsonType(value) === 'object' && value !== null) {
    const obj = value as Record<string, unknown>;

    // required
    if (Array.isArray(schema.required)) {
      for (const key of schema.required) {
        if (typeof key !== 'string') continue;
        if (!Object.prototype.hasOwnProperty.call(obj, key)) {
          return fail(`missing required field "${key}"`, path);
        }
      }
    }

    // properties
    const properties = schema.properties as Record<string, SchemaNode> | undefined;
    if (properties) {
      for (const [key, subSchema] of Object.entries(properties)) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          const result = check(subSchema, obj[key], [...path, key]);
          if (!result.ok) return result;
        }
      }
    }

    // additionalProperties: only enforce when explicitly false. Our schema
    // sets additionalProperties:true everywhere top-level, so this is mostly
    // a guard for future tightening.
    if (schema.additionalProperties === false && properties) {
      for (const key of Object.keys(obj)) {
        if (!Object.prototype.hasOwnProperty.call(properties, key)) {
          return fail(`unexpected field "${key}"`, path);
        }
      }
    }
  }

  // oneOf — discriminator-aware. If a strict-const branch matches the
  // input's `type` field, only that branch is allowed: failures there
  // do NOT fall through to the generic permissive variant. This mirrors
  // the Rust serde(tag = "type") enum dispatch on the consumer side and
  // closes a hole where malformed strict events (missing required fields,
  // wrong-type values) were silently rescued by the generic branch.
  //
  // When no strict branch's discriminator matches, the generic branch is
  // tried; its own `type` enum decides whether the event is a known generic
  // discriminator or an unknown event class that must be rejected.
  if (Array.isArray(schema.oneOf)) {
    const branches = schema.oneOf as SchemaNode[];

    // Look for a strict-const branch whose const matches the input's type.
    for (const rawBranch of branches) {
      const branch = deref(rawBranch);
      if (branchIsStrictConst(branch) && branchTypeMatches(branch, value)) {
        // Locked in: this is the only branch allowed to match. Return its
        // result verbatim — no fallthrough to other branches (incl. generic).
        const result = check(branch, value, path);
        return result.ok ? ok() : result;
      }
    }

    // No strict branch claimed this `type`. Try every remaining branch
    // (non-strict-const, i.e. the generic variant). Pick the deepest-path
    // failure as the best diagnostic if none match.
    let bestFail: ValidationResult | null = null;
    for (const rawBranch of branches) {
      const branch = deref(rawBranch);
      if (branchIsStrictConst(branch)) continue; // already eliminated above
      const result = check(branch, value, path);
      if (result.ok) return ok();
      if (
        bestFail === null ||
        (!bestFail.ok && !result.ok && result.path.length > bestFail.path.length)
      ) {
        bestFail = result;
      }
    }
    if (bestFail !== null) return bestFail;
    return fail('no oneOf branch matched', path);
  }

  return ok();
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Validate an inbound event (the JSONL shape Bridge.enqueue is about to
 *  serialize). Returns ok:false with a path when invalid. */
export function validate(event: unknown): ValidationResult {
  if (jsonType(event) !== 'object' || event === null) {
    return fail(`expected object, got ${jsonType(event)}`, []);
  }
  return check(SCHEMA, event, []);
}

/** Validate an outbound control command — separate top-level oneOf at
 *  schema.outboundCommands. Used by control-channel emitters. */
export function validateCommand(command: unknown): ValidationResult {
  const out = SCHEMA.outboundCommands as SchemaNode | undefined;
  if (out === undefined) return fail('outboundCommands schema missing', []);
  if (jsonType(command) !== 'object' || command === null) {
    return fail(`expected object, got ${jsonType(command)}`, []);
  }
  return check(out, command, []);
}

/** Re-export the loaded schema for callers (tests, tooling). Not for mutation. */
export function schema(): SchemaNode {
  return SCHEMA;
}

File: inspector/src/schema.rs
//! Minimal JSON Schema validator for the Enchanter event-bus wire format.
//!
//! Mirrors `src/observability/schema.ts` on the TS side. Loads
//! `docs/event-schema.json` at compile time via `include_str!`, parses it
//! once into a `serde_json::Value`, and offers `validate(value)` — the same
//! shape as the TS `validate(event)` function so the two sides log
//! consistent error paths.
//!
//! Hand-rolled — supports exactly the JSON Schema keywords this project's
//! schema uses: `type`, `properties`, `required`, `oneOf`, `enum`, `const`,
//! `minimum`, `$ref`, `additionalProperties`. Not a full draft 2020-12
//! validator. Intentional: "no new top-level dep" on either side of the wire.

use std::sync::OnceLock;

use serde_json::Value;

/// Embedded canonical schema. Path is relative to *this source file*
/// (`inspector/src/schema.rs` → `../../docs/event-schema.json`).
const SCHEMA_TEXT: &str = include_str!("../../docs/event-schema.json");

/// Validation error: human-readable reason + JSON Pointer path
/// (slash-joined) so logs like `at /payload/cost_usd` localize the failure.
#[derive(Debug, Clone)]
pub struct SchemaError {
    pub reason: String,
    pub path: Vec<String>,
}

impl SchemaError {
    pub fn pointer(&self) -> String {
        if self.path.is_empty() {
            "/".to_string()
        } else {
            format!("/{}", self.path.join("/"))
        }
    }
}

impl std::fmt::Display for SchemaError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} at {}", self.reason, self.pointer())
    }
}

impl std::error::Error for SchemaError {}

fn schema_root() -> &'static Value {
    static ROOT: OnceLock<Value> = OnceLock::new();
    ROOT.get_or_init(|| {
        serde_json::from_str::<Value>(SCHEMA_TEXT)
            .expect("event-schema.json failed to parse — this is a build-time invariant")
    })
}

fn definitions() -> &'static serde_json::Map<String, Value> {
    static DEFS: OnceLock<serde_json::Map<String, Value>> = OnceLock::new();
    DEFS.get_or_init(|| {
        schema_root()
            .get("definitions")
            .and_then(|v| v.as_object())
            .cloned()
            .unwrap_or_default()
    })
}

/// Resolve a `$ref` once. Returns the referenced node if the ref matches
/// `#/definitions/<name>` and the name exists; otherwise returns `node` as-is.
fn deref<'a>(node: &'a Value) -> &'a Value {
    if let Some(Value::String(r)) = node.get("$ref") {
        if let Some(name) = r.strip_prefix("#/definitions/") {
            if let Some(target) = definitions().get(name) {
                // definitions() is &'static, which outlives any 'a — the
                // narrower-lifetime coercion is sound.
                return target;
            }
        }
    }
    node
}

fn json_type(value: &Value) -> &'static str {
    match value {
        Value::Null => "null",
        Value::Bool(_) => "boolean",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}

fn type_matches(declared: &str, actual: &str) -> bool {
    if declared == "integer" {
        return actual == "number";
    }
    declared == actual
}

fn fail(reason: impl Into<String>, path: &[String]) -> SchemaError {
    SchemaError {
        reason: reason.into(),
        path: path.to_vec(),
    }
}

fn check(node: &Value, value: &Value, path: &mut Vec<String>) -> Result<(), SchemaError> {
    let schema = deref(node);

    // const
    if let Some(expected) = schema.get("const") {
        if value != expected {
            return Err(fail(
                format!(
                    "expected const {}, got {}",
                    expected,
                    truncate_for_msg(value),
                ),
                path,
            ));
        }
    }

    // enum
    if let Some(Value::Array(variants)) = schema.get("enum") {
        if !variants.iter().any(|v| v == value) {
            return Err(fail(
                format!("value {} not in enum", truncate_for_msg(value)),
                path,
            ));
        }
    }

    // type
    if let Some(Value::String(declared)) = schema.get("type") {
        let actual = json_type(value);
        if !type_matches(declared, actual) {
            return Err(fail(
                format!("expected type {}, got {}", declared, actual),
                path,
            ));
        }
    }

    // minimum
    if let Some(min) = schema.get("minimum").and_then(Value::as_f64) {
        if let Some(n) = value.as_f64() {
            if n < min {
                return Err(fail(format!("value {} below minimum {}", n, min), path));
            }
        }
    }

    // object: required, properties, additionalProperties
    if let Value::Object(obj) = value {
        if let Some(Value::Array(required)) = schema.get("required") {
            for req in required {
                if let Some(key) = req.as_str() {
                    if !obj.contains_key(key) {
                        return Err(fail(format!("missing required field \"{}\"", key), path));
                    }
                }
            }
        }

        let properties = schema
            .get("properties")
            .and_then(Value::as_object);

        if let Some(props) = properties {
            for (key, sub_schema) in props {
                if let Some(child) = obj.get(key) {
                    path.push(key.clone());
                    let result = check(sub_schema, child, path);
                    path.pop();
                    result?;
                }
            }
        }

        if let Some(Value::Bool(false)) = schema.get("additionalProperties") {
            if let Some(props) = properties {
                for key in obj.keys() {
                    if !props.contains_key(key) {
                        return Err(fail(format!("unexpected field \"{}\"", key), path));
                    }
                }
            }
        }
    }

    // oneOf — first match wins. If none match, prefer the failure from the
    // branch whose `type` discriminator matched the input — that's the
    // "intended" branch. Fall back to deepest-path otherwise.
    if let Some(Value::Array(branches)) = schema.get("oneOf") {
        let mut best: Option<(SchemaError, bool)> = None;
        for branch in branches {
            match check(branch, value, path) {
                Ok(()) => return Ok(()),
                Err(e) => {
                    let intended = branch_type_matches(deref(branch), value);
                    let replace = match &best {
                        None => true,
                        Some((b, b_intended)) => {
                            if intended && !*b_intended {
                                true
                            } else if intended == *b_intended {
                                e.path.len() > b.path.len()
                            } else {
                                false
                            }
                        }
                    };
                    if replace {
                        best = Some((e, intended));
                    }
                }
            }
        }
        return Err(best
            .map(|(e, _)| e)
            .unwrap_or_else(|| fail("no oneOf branch matched", path)));
    }

    Ok(())
}

/// Heuristic: does this branch's `type` discriminator (a const or an enum on
/// the `type` property) match the `type` field of the candidate object?
fn branch_type_matches(branch: &Value, value: &Value) -> bool {
    let Some(obj) = value.as_object() else {
        return false;
    };
    let Some(t) = obj.get("type").and_then(Value::as_str) else {
        return false;
    };
    let Some(properties) = branch.get("properties").and_then(Value::as_object) else {
        return false;
    };
    let Some(type_schema) = properties.get("type") else {
        return false;
    };
    if let Some(Value::String(c)) = type_schema.get("const") {
        return c == t;
    }
    if let Some(Value::Array(variants)) = type_schema.get("enum") {
        return variants.iter().any(|v| v.as_str() == Some(t));
    }
    false
}

fn truncate_for_msg(value: &Value) -> String {
    let s = value.to_string();
    if s.len() <= 80 {
        s
    } else {
        format!("{}…", &s[..80])
    }
}

/// Validate an inbound event (the JSON object the Rust transport just
/// parsed). Mirrors the TS-side `validate(event)`.
pub fn validate(value: &Value) -> Result<(), SchemaError> {
    if !value.is_object() {
        return Err(fail(
            format!("expected object, got {}", json_type(value)),
            &[],
        ));
    }
    let mut path: Vec<String> = Vec::new();
    check(schema_root(), value, &mut path)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn happy_path_runtime_metrics() {
        let v = json!({
            "type": "runtime.metrics",
            "time": 1.0,
            "open_sessions": 3,
            "ongoing_tasks": 5,
            "queued_tasks": 2,
            "blocked_tasks": 1,
            "code_written_lifetime_loc": 0,
            "code_modified_lifetime_loc": 0,
            "files_created_lifetime": 0,
            "files_modified_lifetime": 0,
            "tool_calls_lifetime": 0,
            "prs_created_lifetime": 0,
            "tests_run_lifetime": 0,
            "tests_passed_rate": 0.94,
            "total_spend_lifetime": 0.0
        });
        validate(&v).expect("should validate");
    }

    #[test]
    fn rejects_missing_required() {
        let v = json!({"type": "tool.call", "time": 1.0, "payload": {}});
        assert!(validate(&v).is_err());
    }

    #[test]
    fn rejects_bad_enum() {
        let v = json!({
            "type": "hydra.veto", "time": 1.0, "policy": "p", "reason": "r",
            "action": "a", "severity": "fatal", "payload": null
        });
        assert!(validate(&v).is_err());
    }

    #[test]
    fn rejects_unknown_discriminator() {
        let v = json!({"type": "totally.unknown", "time": 1.0});
        assert!(validate(&v).is_err());
    }
}
